### PR TITLE
Add job status listener [HZ-1646]

### PR DESCRIPTION
**Pull request by burakgok**
For hazelcast/hazelcast#23193 and hazelcast/hazelcast-enterprise#5707

### CODE REVIEW

**Comment by mdumandag** [comment]
Hi, just a heads up: We have updated the code generator to use 2023 as the copyright year. So, don't forget to rebase your branch with master and regenerate codecs and test files so that they have the correct copyright year in the Hazelcast PR.

File: protocol-definitions/Jet.yaml [hunk 1]
```diff
@@ -301,3 +301,107 @@ methods:
...
+      - name: Job
+        since: 2.6
+        params:
+          - name: jobId
+            type: long
+            nullable: false
+            since: 2.6
+            doc: |
+              Id of job.
+          - name: oldStatus
```
Comment (frant-hartm) [inline protocol-definitions/Jet.yaml:343]: I would prefer `previousStatus`

**Review by mdumandag — DISMISSED** [review]

**Review by mdumandag — APPROVED** [review]

**Review by frant-hartm — DISMISSED** [review]

**Review by ihsandemir — DISMISSED** [review]

File: protocol-definitions/Jet.yaml [hunk 2]
```diff
@@ -416,3 +416,106 @@ methods:
...
+              of SUSPEND, RESUME, RESTART and CANCEL; if there is a failure, indicates
+              the cause; otherwise, null.
+          - name: userRequested
+            type: boolean
+            nullable: false
+            since: 2.6
+            doc: |
+              Indicates whether the event is generated by the user via
+              {@code Job.suspend()}, {@code Job.resume()}, {@code Job.restart()} or
+              {@code Job.cancel()}.
```
Comment (k-jamroz) [inline protocol-definitions/Jet.yaml:491]: + `cancelAndExportSnapshot()`

File: protocol-definitions/Jet.yaml [hunk 3]
```diff
@@ -416,3 +416,106 @@ methods:
...
+            type: int
+            nullable: false
+            since: 2.6
+            doc: |
+              NOT_RUNNING(0)
+              STARTING(1)
+              RUNNING(2)
+              SUSPENDED(3)
+              SUSPENDED_EXPORTING_SNAPSHOT(4)
+              FAILED(6)
```
Comment (k-jamroz) [inline protocol-definitions/Jet.yaml:468]: FAILED and COMPLETED are terminal states. they cannot be `previousStatus`

File: protocol-definitions/Jet.yaml [hunk 4]
```diff
@@ -416,3 +416,106 @@ methods:
...
+          type: boolean
+          nullable: false
+          since: 2.6
+          doc: |
+            If true fires events that originated from this node only, otherwise fires all events.
+    response:
+      params:
+        - name: response
+          type: UUID
+          nullable: false
```
Comment (k-jamroz) [inline protocol-definitions/Jet.yaml:444]: in Java generated code UUID is always nullable (it is one of the "magic" types)

File: protocol-definitions/Jet.yaml [hunk 5]
```diff
@@ -416,3 +416,106 @@ methods:
...
+          since: 2.6
+          doc: |
+            A unique string which is used as a key to remove the listener.
```
Comment (k-jamroz) [inline protocol-definitions/Jet.yaml:447]: UUID is not encoded as string but as bytes

**Review by k-jamroz — APPROVED** [review]

**Review by mdumandag — APPROVED** [review]

File: protocol-definitions/Jet.yaml [hunk 6]
```diff
@@ -416,3 +416,110 @@ methods:
...
+        since: 2.6
+        params:
+          - name: jobId
+            type: long
+            nullable: false
+            since: 2.6
+            doc: |
+              ID of job.
+          - name: previousStatus
+            type: int
```
Comment (ihsandemir) [inline protocol-definitions/Jet.yaml:459]: why such a large number(4-bytes) for an enum?  why not `byte`?

Comment (ihsandemir) [reply protocol-definitions/Jet.yaml:459]: Another unfortunate choice of types.

**Review by ihsandemir — APPROVED** [review]

**Pull request merged**